### PR TITLE
GH-30891: [C++] The C++ API for writing datasets could be improved

### PR DESCRIPTION
**Pull request by vibhatha**
### Rationale for this change

This PR addresses the issues mentioned [here](https://github.com/apache/arrow/issues/30891#issue-1528196430)

### What changes are included in this PR?

The test cases are updated with no-arg constructor update on `FileSystemDatasetWriteOptions` along with the default values.

### Are these changes tested?

TODO

### Are there any user-facing changes?

Yes, the examples have been updated.

WIP: based on CI status, other changes will be made.
* Closes: #30891

### CODE REVIEW

**Comment by kou** [comment]
GLib/Ruby parts are updated.

**Comment by vibhatha** [comment]
@kou one of the CIs related Ruby is failing. It seems like an unrelated issue, but should we re-run it?

**Comment by kou** [comment]
The failed job was solved by https://github.com/apache/arrow/pull/33815 .
So we can fix it by rebasing on master.

**Comment by vibhatha** [comment]
> The failed job was solved by #33815 .
> So we can fix it by rebasing on master.

Great.

**Comment by vibhatha** [comment]
cc @westonpace could you please take a look?

**Review by westonpace — CHANGES_REQUESTED** [review]
We should be careful about changing public APIs if we can avoid it.  I'm not sure why the prototype for `FileSystemDataset::Write` needed to change though it's possible I'm missing something.

File: cpp/src/arrow/dataset/file_base.h [hunk 1]
```diff
@@ -450,20 +465,25 @@ struct ARROW_DS_EXPORT FileSystemDatasetWriteOptions {
   };
 
   const std::shared_ptr<FileFormat>& format() const {
-    return file_write_options->format();
+    return file_write_options_->format();
   }
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.h:469]: I think we cannot have `format` as a public attribute and a method with `format()`, right? That's why I kept using the method. With this new modification, may be we can get rid of the `format()` function

File: cpp/src/arrow/dataset/file_base.h [hunk 2]
```diff
@@ -280,7 +281,7 @@ class ARROW_DS_EXPORT FileSystemDataset : public Dataset {
       std::shared_ptr<Partitioning> partitioning = NULLPTR);
 
   /// \brief Write a dataset.
-  static Status Write(const FileSystemDatasetWriteOptions& write_options,
+  static Status Write(const std::shared_ptr<FileSystemDatasetWriteOptions>& write_options,
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.h]: From the offline discussion, decided to revert this change.

**Comment by vibhatha** [comment]
> @vibhatha Can you close this out if you're not going to wrap it up?

I won’t be able to wrap it up in the coming weeks. I will close this for now and reopen.

**Pull request closed without merging**